### PR TITLE
Add tests, tweak implementation

### LINKED ISSUE #930: Cartographer stamps RunTemplates continuously when referenced by multiple Runnables
When a RunTemplate is referenced by multiple Runnables, Cartographer will endlessly stamp the template alternately for each Runnable.

Reproduce with the following (courtesy of @waciumawanjohi):
```
---
apiVersion: tekton.dev/v1beta1
kind: ClusterTask
metadata:
  name: passing-cluster-task
spec:
  description: |-
    A task that passes
  steps:
    - name: solo-step
      image: busybox
      script: |
        exit 0

---
apiVersion: carto.run/v1alpha1
kind: ClusterRunTemplate
metadata:
  name: passing-template
spec:
  template:
    apiVersion: tekton.dev/v1beta1
    kind: TaskRun
    metadata:
      generateName: passing-
    spec:
      taskRef:
        name: passing-cluster-task
        kind: ClusterTask


---
apiVersion: carto.run/v1alpha1
kind: Runnable
metadata:
  name: passing
spec:
  runTemplateRef:
    name: passing-template

---
apiVersion: carto.run/v1alpha1
kind: Runnable
metadata:
  name: passing-2
spec:
  runTemplateRef:
    name: passing-template
```

### PR DESCRIPTION
**Pull request by waciumawanjohi**
- Adds a regression test
- Implements discriminant creation using O(nlogn) algo rather than O(n^2)

## Release Note

Runnable stamped resources now also key by their labels in the cache, ensuring stamped resources from different Runnables sharing the same RunTemplate do not cause endless resources to be stamped [Fix #930]

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [ ] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- [ ] Filled in the [Release Note](#Release-Note) section above 
- [ ] Modified the docs to match changes

### CODE REVIEW

File: pkg/repository/repository.go [hunk 1]
```diff
@@ -501,16 +501,16 @@ func (r *repository) GetScheme() *runtime.Scheme {
 
 func buildOwnerDiscriminant(labels map[string]string) string {
 	var discriminantComponents []string
-	for key, value := range labels {
-		discriminantComponents = insertSorted(discriminantComponents, fmt.Sprintf("{%s:%s}", key, value))
+
```
Comment (idoru) [inline pkg/repository/repository.go:504]: Good spot. I'm also currently working on another PR which unifies cache key creation at a different layer (the addition of the discriminant has stretched the original design where Cache was aware of how to make cache keys) and would rather work this into there, if the final solution requires sorting at all. My view was either #930 is critical enough that we can merge #931 and #932 first or it could wait for the refactoring.

File: tests/integration/runnable/runnable_test.go [hunk 2]
```diff
@@ -259,6 +260,55 @@ var _ = Describe("Stamping a resource on Runnable Creation", func() {
 					}))
 				})
 			})
+
+			Context("and a second Runnable matches the RunTemplateRef", func() {
```
Comment (idoru) [inline tests/integration/runnable/runnable_test.go:264]: I'm not a fan of this style of regression test, as it's key assertion is blindly waiting for something to not happen. As such, it either has to hope it has waited long enough (else you get false positives), or always has to wait longer than necessary, protracting test execution time.

To be clear, I'm not saying this test is worthless, but my calculus was: If I know we already have tests like this for a single object only being stamped once in here (per input change) which implies the Cache, then is such an integration test much akin to a corner case test for a `Repository -> Cache` bug? If so, what value does such an integration test add that makes it worth paying the time / uncertainty cost?

Interestingly, I think if we are destined to ultimately use ReconcilerTestSuite, then this and similarly structured tests would be converted to forms that reduce the cost (increased test execution time) and eliminate uncertainty.

File: pkg/repository/repository.go (hunk 1 above)
Comment (waciumawanjohi) [reply pkg/repository/repository.go:504]: Assuming [this doc](https://confluence.eng.vmware.com/pages/viewpage.action?pageId=789919969) is still accurate, we want to get the fix into TAP by Friday.

File: tests/integration/runnable/runnable_test.go (hunk 2 above)
Comment (waciumawanjohi) [reply tests/integration/runnable/runnable_test.go:264]: This test unquestionably adds 2 seconds to the integration tests.

> is such an integration test much akin to a corner case test for a Repository -> Cache bug

This is undoubtedly a regression test. There was a bug, it effected customers.

As you point out, this test is similar to the assertion block above it for stamping out a single object. We can reduce the 2 second delay by combining this test with the other (and thus have one 2 second wait rather than two).

File: pkg/repository/repository.go (hunk 1 above)
Comment (idoru) [reply pkg/repository/repository.go:504]: That looks correct. I agree this PR improves on #931 / #932 and in essence I would approve, but since it's based on these PRs which I worked on and they've not been approved yet, it didn't quite feel legitimate for me to do so.

**Review by idoru — APPROVED** [review]
